### ISSUE TITLE
[BUG] Build RAFT tests with `-Wall -Werror`

### ISSUE THREAD

**Issue by trxcllnt** [rapidsai/raft#225]
As a header-only library, warnings in RAFT can cause errors in consuming libs that compile with `-Werror`. Building tests with `-Wall -Werror` [like RMM does](https://github.com/rapidsai/rmm/blob/7768d4d0faa3401e9dd455db47125370be63b983/tests/CMakeLists.txt#L56-L57) should catch most problems before they're released.

For example, see this error in a RAFT header when compiling cuML:
```
raft/cpp/include/raft/sparse/hierarchy/detail/mst.cuh: In instantiation of 'void raft::hierarchy::detail::build_sorted_mst(const raft::handle_t&, const value_t*, const value_idx*, const value_idx*, const value_t*, size_t, size_t, rmm::device_uvector<value_idx>&, rmm::device_uvector<value_idx>&, rmm::device_uvector<value_t>&, size_t, raft::distance::DistanceType, int) [with value_idx = int; value_t = float; size_t = long unsigned int]':
raft/cpp/include/raft/sparse/hierarchy/single_linkage.hpp:80:47:   required from 'void raft::hierarchy::single_linkage(const raft::handle_t&, const value_t*, size_t, size_t, raft::distance::DistanceType, raft::hierarchy::linkage_output<value_idx, value_t>*, int, size_t) [with value_idx = int; value_t = float; raft::hierarchy::LinkageDistance dist_type = raft::hierarchy::PAIRWISE; size_t = long unsigned int]'
cuml/cpp/src/hierarchy/linkage.cu:32:117:   required from here
raft/cpp/include/raft/sparse/hierarchy/detail/mst.cuh:204:410: warning: format '%d' expects argument of type 'int', but argument 4 has type 'size_t' {aka 'long unsigned int'} [-Wformat=]
```

**Comment by dantegd** [comment]
PR #187 adds the cmake side to enable this easily, was going to push it in the same PR but (as usual) once I fix what it looked like a single warning, more started popping up, so it probably needs to be a follow up.

**Comment by dantegd** [comment]
Closed nu #299
